### DISCUSSION TITLE
How to customize password strength rules?

### DISCUSSION THREAD

**Discussion by ronnylam1210** [serenity-is/Serenity#7182] [Help] [answered]
I want to modify the default password strength rules. How should I modify it?
Thank you!

**Comment by LSNicholls** [comment]
there is a class called  PasswordStrengthValidator, and this is set up in Startup.cs:

  `    services.AddSingleton<IPasswordStrengthValidator, PasswordStrengthValidator>();`

... so you have to write your own class that implements the interface IPasswordStrengthValidator and substitute for the above line in Startup.cs to call your class.

**Comment by ronnylam1210** [comment]
@LSNicholls I'll try this, thanks for your reply!

**Comment by LSNicholls** [comment]
@ronnylam1210 actually it may be easier than that, if all you want to do is make some changes according to the abilities of the default PasswordStrengthValidator class.  Have a look at https://github.com/serenity-is/common-features/blob/master/src/Serenity.Extensions/Modules/Membership/MembershipSettings.cs.  If you just need to tweak those settings, you just change appsettings.json and you're done, right?

Sorry, I didn't think of that earlier!

**Comment by ronnylam1210** [comment]
@LSNicholls Yes, it seems that modifying MembershipSettings can achieve the function I want, but this is in Serenity.Extensions. How should I modify or create a new setting file to achieve customized rules? Thanks for your reply!

**Comment by LSNicholls** [comment, marked as answer]
> How should I modify or create a new setting file to achieve customized rules?

you don't have to create a new file.  You just need to put a Membership section into appsettings.json, as I said.  Right?  for example:

```
,
  "Membership": {
    "MinPasswordLength": 20
  }
```

... use the MembershipSettings.cs file to understand what you can put in that section.

**Comment by ronnylam1210** [comment]
When I put custom membership rules in appsettings, it worked!
But after clicking change password, a blank alert popped up on the screen, and the password was not successfully changed...any idea?
![image](https://github.com/serenity-is/Serenity/assets/33535307/2269ade2-fef5-4bf7-b80b-eb0815db2465)

![image](https://github.com/serenity-is/Serenity/assets/33535307/9c5d2f2e-70f0-45d1-9075-9de796774a79)

**Comment by LSNicholls** [comment]
>.any idea?

@ronnylam1210 , yes.  Thank you for providing the error information.  It seems that your answer is right there.

![image](https://github.com/serenity-is/Serenity/assets/88156108/49e463e0-bc54-4447-ae65-4d27431de331)

Did you comment out the line that sets up the default PasswordStrengthValidator in Startup.cs, in preparation for adding one of your own, and then forget to un-comment it when you settled on the approach that only needs you to put a Membership section in appsettings.json?

**Comment by ronnylam1210** [comment]
@LSNicholls The function can be used normally, thanks you for your help!!
